File: Projects/Graffer/0.2/grafer/ui/DrawArea.py
'''	
	This file is part of Camarillo.
	
	Copyright (C) 2008	Frederic-Gerald Morcos <fred.morcos@gmail.com>
	
	Camarillo is free software: you can redistribute it and/or modify
	it under the terms of the GNU General Public License as published by
	the Free Software Foundation, either version 3 of the License, or
	(at your option) any later version.

	Camarillo is distributed in the hope that it will be useful,
	but WITHOUT ANY WARRANTY; without even the implied warranty of
	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.	See the
	GNU General Public License for more details.

	You should have received a copy of the GNU General Public License
	along with Camarillo.	If not, see <http://www.gnu.org/licenses/>.
'''

from gtk import ScrolledWindow, HRuler, VRuler, Table, \
				POLICY_ALWAYS, SHRINK, FILL, STATE_NORMAL
from gtk.gdk import Color, \
					POINTER_MOTION_MASK
from goocanvas import Canvas

class DrawArea (ScrolledWindow):
	def __init__(self):
		ScrolledWindow.__init__(self)
		
		self.set_policy(POLICY_ALWAYS, POLICY_ALWAYS)
		
		self.table = Table(3, 3)
		self.hRuler = HRuler()
		self.vRuler = VRuler()
		self.canvas = Canvas()
		
		self.canvas.connect('size-allocate', self.area_allocate, None)
		self.canvas.connect('motion-notify-event', self.area_motion, None)
		
		self.table.attach(self.hRuler, 1, 2, 0, 1, SHRINK | FILL, SHRINK | FILL)
		self.table.attach(self.vRuler, 0, 1, 1, 2, SHRINK | FILL, SHRINK | FILL)
		self.table.attach(self.canvas, 1, 2, 1, 2)
		
		self.add_with_viewport(self.table)

	def area_allocate(self, widget, allocation, data):
		self.hRuler.set_range(
							0, self.canvas.allocation[2], 
							10, self.canvas.allocation[2])
		self.vRuler.set_range(
							0, self.canvas.allocation[3], 
							10, self.canvas.allocation[3])

	def area_motion(self, widget, event, data):
		self.hRuler.set_range(
							0, self.canvas.allocation[2], 
							event.x, self.canvas.allocation[2])
		self.vRuler.set_range(
							0, self.canvas.allocation[3], 
							event.y, self.canvas.allocation[3])

File: Projects/GASPaint/other/Cairo/src/drawarea.c
/*
 * This file is part of gaspaint
 * 
 * Copyright (C) 2007-2009  Frederic-Gerald Morcos
 * 
 * gaspaint is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 * 
 * gaspaint is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with gaspaint.  If not, see <http://www.gnu.org/licenses/>.
 */

#include <gtk/gtk.h>
#include <string.h>

#include "drawarea.h"
#include "public.h"
#include "config.h"
#include "drawing.h"
#include "event.h"

/* our drawing area */

/* is called when the drawing area is clicked,
 * manages the various clicks for the various tools
 */
void draw_area_clicked (GtkWidget *widget, GdkEventButton *event)
{		
	gint x = event->x, y = event->y;
	
	if (num_clicks == 0)
	{
		num_clicks++;
		current_x0 = x;
		current_y0 = y;
		
		if (current_tool == FILL)
		{
			char command [25];
			num_clicks = 0;
			sprintf (command, "fill %d %d", current_x0, current_y0);
			add_event (command);
			draw_area_expose ();
		}
	}
	else if (num_clicks == 1)
	{
		num_clicks = 0;
		current_x1 = x;
		current_y1 = y;
		
		char command [25];		/* 3*4 digits + 4 (spaces) + 4 chars (command name) + \0 */
		
		if (current_tool == LINE)
		{
			sprintf (command, "line %d %d %d %d", current_x0, current_y0, current_x1, current_y1);
			add_event (command);
			draw_area_expose ();
		}
		else if (current_tool == RECT)
		{
			sprintf (command, "rect %d %d %d %d", current_x0, current_y0, current_x1, current_y1);
			add_event (command);
			draw_area_expose ();
		}
		else if (current_tool == CIRC)
		{
			sprintf (command, "circ %d %d %d %d", current_x0, current_y0, current_x1, current_y1);
			add_event (command);
			draw_area_expose ();
		}
		else if (current_tool == POLY)
		{
			num_clicks = 1;
			sprintf (command, "line %d %d %d %d", current_x0, current_y0, current_x1, current_y1);
			current_x0 = current_x1;
			current_y0 = current_y1;
			
			add_event (command);
			draw_area_expose ();
		}
	}
	else
	{
		current_x0 = -1;
		current_y0 = -1;
		current_x1 = -1;
		current_y1 = -1;
		num_clicks = 0;
	}
}

/* called when the mouse is moving over the drawing area */
void draw_area_motion (GtkWidget *widget, GdkEventMotion *event, GtkWidget *label)
{
	gint x = event->x, y = event->y;
	char pos [20];
	
	sprintf (pos, "X: %d, Y: %d", x, y);
	gtk_label_set_label (GTK_LABEL (label), pos); 
}

